service: Simplify logic of svc ID allocation

brb opened the pull request:
This PR changes the logic of the service ID allocation which allows to simplify a code of the callers.

Previously, if a requested service ID had been allocated to another service, the allocation request silently allocated a new service ID. This behavior required additional checks at the caller side to validate whether the allocated ID and the requested ID were the same. This could lead to a service ID leak if the caller forgot to release the ID if the IDs were not the same.

The change is to return an error if a requested ID is already taken. This allows us to get rid of some boilerplate code when creating a service.

Reviewable per commit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9213)
<!-- Reviewable:end -->

brb committed:
lbmap: Remove dead code

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Remove redundant check for SVC ID

The service ID is acquired only in k8s_watcher after receiving
a service update event from service_cache, therefore it's
always null before the acquisition.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Return error if ID has been already taken

This commit changes the logic of global and local service ID allocator.

Previously, if a requested service ID had been allocated to another
service, the allocation request silently allocated a new service ID.
This behavior required additional checks at the caller side to validate
whether the allocated ID and the requested ID were the same.

The change will simplify the caller code.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Remove redundant checks for acquired service ID

The ID allocator will return an error if the required ID is allocated
to another service.

Also, this fixes the svc ID leak, as the ID was not released before
returning the error.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Move service ID acquisition to svcAdd()

Previously, both HTTP PUT service/{id} and k8s.UpdateService
handlers were allocating svc IDs before calling d.svcAdd().

To DRY, we move the ID allocation call to d.svcAdd().

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Release service ID upon failure

This should prevent from leaking svc IDs when a service insertion
failed after the ID has been acquired.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on daemon/k8s_watcher.go:
```diff
@@ -1256,25 +1256,23 @@ func (d *Daemon) addK8sSVCs(svcID k8s.ServiceID, svc *k8s.Service, endpoints *k8
-				logfields.Object:      logfields.Repr(svc),
-			}).Debug("Got feAddr ID for service")
-			fePort.ID = loadbalancer.ServiceID(feAddrID.ID)
+		feAddr := loadbalancer.NewL3n4Addr(fePort.Protocol, svc.FrontendIP, fePort.Port)
```
what about when a service is added through the cilium-agent's CLI?

brb replied:
You can specify it via the `--id` param. E.g. `cilium service update ... --id 1234`.

aanm commented on daemon/loadbalancer.go:
```diff
@@ -130,7 +140,8 @@ func (d *Daemon) svcAdd(
+		id, err = service.LookupBackendID(be.L3n4Addr)
 		if err != nil {
 			scopedLog.WithField(logfields.BackendName, be.L3n4Addr).WithError(err).
 				Warning("Unable to lookup backend ID")
```
Why do we continue if err != nil? If err is not nil for index 0 but is nil for index 1, why is that ok?

brb replied:
This is unrelated to my PR, and this is not critical, as the retrieved IDs are used only for `cilium service list`.

Anyway, I have in my LB refactoring TODO to get rid of this.

aanm replied:
@brb it is related with your PR because the `err` in line 144 can be the same of line `99`

brb replied:
Oops, nice catch. Fixed.

brb committed:
daemon: Add check for acquired service ID

Check whether acquired service ID is the same as the requested one.
Return error if it's not the same to prevent users from accidentally
updating wrong services.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
